packages/hooks/src/useInputDecimalFormatter.ts

```diff
@@ -0,0 +1,72 @@
+import { useCallback, useEffect, useState } from 'react';
+
+type TOptions = {
+    fraction_digits?: number;
+    with_sign?: boolean;
+};
+
+const useInputDecimalFormatter = (initial?: number, options?: TOptions) => {
+    const [value, setValue] = useState('');
+    const { fraction_digits = 2, with_sign = false } = options || {};
+
+    const onChange = useCallback(
+        (e: DeepPartial<React.ChangeEvent<HTMLInputElement>> | React.ChangeEvent<HTMLInputElement>) => {
+            setValue(old_value => {
+                const new_value = e?.target?.value || '';
+                const isEmpty = new_value === '';
+
+                // The field has been cleared, So we return the new value.
+                if (isEmpty) return new_value;
+
+                const text = with_sign ? new_value : new_value.replaceAll(/[+-]/g, '');
+                const inputs = text.split('.');
+
+                // The field contains more than one dot, So we return the old value as only one dot
+                // is allowed.
+                if (inputs.length > 2) return old_value;
+
+                const left = inputs[0];
+                const right = inputs.length > 1 ? inputs[1] : null;
+                const has_right = right !== null && right !== '';
+
+                // The field value is positive or negative sign, So we return the new value without
+                // any calculations.
+                if ((left === '-' || left === '+') && !has_right) return new_value;
+
+                // The field value is 0, So we return the new value without any calculations.
+                if (left === '0' && !has_right) return new_value;
+
+                const is_number = !isNaN(Number(new_value));
+
+                // The input value is not a valid number, So we return the old value.
+                if (!is_number) return old_value;
```

Comment:
the old value would always be an empty string , right? :thinking:

Reply:
unless for instance user tries to type sth like `12345` and then after that try to type 'b', 'c'  ( to have sth like 12345bc ) then it will prevent them to even get to the 'c' and it will remove 'b' by force and then the same thing for 'c' . right?

Reply:
@shayan-deriv Not always, The old value is the last value in the input, So if the new value is something invalid and we don't like it we just keep the old one (which means not accepting the new value)

You can check the test file for it to see how it behaves, Let me know if you need a call to discuss it 🙇🏻

Reply:
got it. thanks